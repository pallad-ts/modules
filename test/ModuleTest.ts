import {DummyModule} from "./DummyModule";
import * as sinon from 'sinon';

describe('Module', () => {

    let module: DummyModule;
    let actions: DummyModule.ActionDefinition[];
    let onInit: sinon.SinonStub;

    const ACTION_NAME1 = 'action1';
    const ACTION_NAME2 = 'action2';
    const CONTEXT = {foo: 'bar'};

    beforeEach(() => {
        actions = [
            {name: ACTION_NAME1, action: sinon.stub()},
            {name: ACTION_NAME2, action: sinon.stub()}
        ];
        onInit = sinon.stub();
        module = new DummyModule('foo', actions, onInit);
    });

    describe('checking whether action exists', () => {
        it('exists', () => {
            expect(module.hasAction(ACTION_NAME1))
                .toBe(true);
        });

        it('does not exists', () => {
            expect(module.hasAction('some-random'))
                .toBe(false);
        });
    });

    describe('registering action', () => {
        it('fails to register action with same name', () => {
            expect(() => {
                // tslint:disable-next-line:no-unused-expression
                new DummyModule('foo', [
                    {name: 'foo', action: sinon.stub()},
                    {name: 'foo', action: sinon.stub()}
                ])
            })
                .toThrowErrorMatchingSnapshot();
        })
    });

    describe('running action', () => {
        it('calls action if exists', () => {
            module.runAction(ACTION_NAME1, CONTEXT);

            sinon.assert.calledWith(actions[0].action as any, sinon.match.same(CONTEXT));
        });

        it('throws an error if action does not exist', () => {
            expect(() => {
                module.runAction('some-random', CONTEXT);
            })
                .toThrowErrorMatchingSnapshot();
        });
    });
});
