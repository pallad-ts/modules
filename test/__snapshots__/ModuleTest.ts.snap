// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module registering action fails to register action with same name 1`] = `"Action \\"foo\\" is already registered in module \\"foo\\""`;

exports[`Module running action throws an error if action does not exist 1`] = `"Action \\"some-random\\" is not registered in module \\"foo\\""`;
